File: apex-trade-bot/src/bybit.js
const axios = require('axios');
const crypto = require('crypto');
const cfg = require('./config');

const BASE = cfg.BYBIT_TESTNET
  ? 'https://api-testnet.bybit.com'
  : 'https://api.bybit.com';

function sign(params) {
  const ts = Date.now().toString();
  const query = Object.keys(params).sort().map(k => `${k}=${params[k]}`).join('&');
  const payload = ts + cfg.BYBIT_API_KEY + '5000' + query;
  const sig = crypto.createHmac('sha256', cfg.BYBIT_API_SECRET).update(payload).digest('hex');
  return { sig, ts };
}

function authHeaders(params) {
  const { sig, ts } = sign(params);
  return {
    'X-BAPI-API-KEY': cfg.BYBIT_API_KEY,
    'X-BAPI-TIMESTAMP': ts,
    'X-BAPI-SIGN': sig,
    'X-BAPI-RECV-WINDOW': '5000',
  };
}

// ─── Surse de date publice (cu fallback automat) ──────────

// Sursa 1: Bybit public klines
async function candlesBybit(symbol, interval, limit) {
  const ivMap = { '1m':'1','3m':'3','5m':'5','15m':'15','30m':'30','1h':'60','4h':'240','1d':'D' };
  const iv = ivMap[interval] || '15';
  const { data } = await axios.get(`${BASE}/v5/market/kline`, {
    params: { category:'spot', symbol, interval: iv, limit },
    headers: { 'User-Agent': 'ApexTradeBot/2.0' },
    timeout: 8000,
  });
  if (data.retCode !== 0) throw new Error('Bybit: ' + data.retMsg);
  return data.result.list.reverse().map(k => ({
    time: parseInt(k[0]), open: parseFloat(k[1]), high: parseFloat(k[2]),
    low: parseFloat(k[3]), close: parseFloat(k[4]), volume: parseFloat(k[5]),
  }));
}

// Sursa 2: Binance public klines
async function candlesBinance(symbol, interval, limit) {
  const { data } = await axios.get('https://api.binance.com/api/v3/klines', {
    params: { symbol, interval, limit },
    headers: { 'User-Agent': 'ApexTradeBot/2.0' },
    timeout: 8000,
  });
  return data.map(k => ({
    time: k[0], open: parseFloat(k[1]), high: parseFloat(k[2]),
    low: parseFloat(k[3]), close: parseFloat(k[4]), volume: parseFloat(k[5]),
  }));
}

// Sursa 3: OKX public candles (fallback final)
async function candlesOKX(symbol, interval, limit) {
  // OKX format: DOGE-USDT, interval: 15m
  const okxSym = symbol.replace('USDT', '-USDT');
  const okxBar = interval; // 1m, 5m, 15m, 1H, 4H, 1D
  const { data } = await axios.get('https://www.okx.com/api/v5/market/candles', {
    params: { instId: okxSym, bar: okxBar, limit },
    headers: { 'User-Agent': 'ApexTradeBot/2.0' },
    timeout: 8000,
  });
  if (data.code !== '0') throw new Error('OKX: ' + data.msg);
  return data.data.reverse().map(k => ({
    time: parseInt(k[0]), open: parseFloat(k[1]), high: parseFloat(k[2]),
    low: parseFloat(k[3]), close: parseFloat(k[4]), volume: parseFloat(k[5]),
  }));
}

// Preț curent — încearcă Bybit → Binance → OKX
async function getPrice(symbol = cfg.SYMBOL) {
  // Bybit
  try {
    const { data } = await axios.get(`${BASE}/v5/market/tickers`, {
      params: { category: 'spot', symbol },
      headers: { 'User-Agent': 'ApexTradeBot/2.0' },
      timeout: 6000,
    });
    const price = data.result?.list?.[0]?.lastPrice;
    if (price) return parseFloat(price);
  } catch {}

  // Binance fallback
  try {
    const { data } = await axios.get('https://api.binance.com/api/v3/ticker/price', {
      params: { symbol },
      headers: { 'User-Agent': 'ApexTradeBot/2.0' },
      timeout: 6000,
    });
    return parseFloat(data.price);
  } catch {}

  // OKX fallback
  const okxSym = symbol.replace('USDT', '-USDT');
  const { data } = await axios.get('https://www.okx.com/api/v5/market/ticker', {
    params: { instId: okxSym },
    headers: { 'User-Agent': 'ApexTradeBot/2.0' },
    timeout: 6000,
  });
  return parseFloat(data.data[0].last);
}

// Lumânări — încearcă Bybit → Binance → OKX
async function getCandles(symbol = cfg.SYMBOL, interval = cfg.TIMEFRAME, limit = cfg.CANDLES) {
  try { return await candlesBybit(symbol, interval, limit); } catch (e) {
    console.warn('[DATA] Bybit klines failed:', e.message, '→ Binance');
  }
  try { return await candlesBinance(symbol, interval, limit); } catch (e) {
    console.warn('[DATA] Binance klines failed:', e.message, '→ OKX');
  }
  return await candlesOKX(symbol, interval, limit);
}

// ─── Balanță (privat Bybit) ───────────────────────────────
async function getBalance(coin = 'USDT') {
  if (cfg.PAPER_TRADING) return cfg.PAPER_BALANCE;
  const params = { accountType: 'UNIFIED', coin };
  const { data } = await axios.get(`${BASE}/v5/account/wallet-balance`, {
    params, headers: authHeaders(params),
  });
  const bal = data.result.list[0]?.coin?.find(c => c.coin === coin);
  return bal ? parseFloat(bal.availableToWithdraw) : 0;
}

// ─── Plasare ordin MARKET (privat Bybit) ─────────────────
async function placeOrder(side, quantity, symbol = cfg.SYMBOL) {
  if (cfg.PAPER_TRADING) {
    console.log(`[PAPER] ${side} ${quantity} ${symbol}`);
    return { orderId: 'PAPER_' + Date.now(), orderStatus: 'Filled' };
  }
  const params = {
    category: 'spot', symbol,
    side: side === 'BUY' ? 'Buy' : 'Sell',
    orderType: 'Market', qty: String(quantity),
  };
  const { data } = await axios.post(`${BASE}/v5/order/create`, params, {
    headers: { ...authHeaders(params), 'Content-Type': 'application/json' },
  });
  return data.result;
}

module.exports = { getPrice, getCandles, getBalance, placeOrder };

File: apex-trade-bot-py/apex/exchanges/bybit.py
"""Bybit Spot connector (API v5) with public-data fallback to Binance/OKX."""
import time
import json
import hmac
import hashlib
import requests
from apex import config as cfg

BASE = "https://api-testnet.bybit.com" if cfg.BYBIT_TESTNET else "https://api.bybit.com"
RECV = "5000"
UA = {"User-Agent": "ApexTradeBot/2.0"}
_IV = {"1m": "1", "3m": "3", "5m": "5", "15m": "15", "30m": "30", "1h": "60", "4h": "240", "1d": "D"}


def _sign(payload):
    return hmac.new(cfg.BYBIT_API_SECRET.encode(), payload.encode(), hashlib.sha256).hexdigest()


def _headers_get(params):
    ts = str(int(time.time() * 1000))
    query = "&".join(f"{k}={params[k]}" for k in sorted(params))
    sig = _sign(ts + cfg.BYBIT_API_KEY + RECV + query)
    return {"X-BAPI-API-KEY": cfg.BYBIT_API_KEY, "X-BAPI-TIMESTAMP": ts,
            "X-BAPI-SIGN": sig, "X-BAPI-RECV-WINDOW": RECV}


def _headers_post(body_str):
    ts = str(int(time.time() * 1000))
    sig = _sign(ts + cfg.BYBIT_API_KEY + RECV + body_str)
    return {"X-BAPI-API-KEY": cfg.BYBIT_API_KEY, "X-BAPI-TIMESTAMP": ts, "X-BAPI-SIGN": sig,
            "X-BAPI-RECV-WINDOW": RECV, "Content-Type": "application/json"}


def get_price(symbol=None):
    symbol = symbol or cfg.SYMBOL
    try:
        r = requests.get(f"{BASE}/v5/market/tickers", params={"category": "spot", "symbol": symbol},
                         headers=UA, timeout=6)
        p = r.json().get("result", {}).get("list", [{}])[0].get("lastPrice")
        if p:
            return float(p)
    except Exception:
        pass
    r = requests.get("https://www.okx.com/api/v5/market/ticker",
                     params={"instId": symbol.replace("USDT", "-USDT")}, headers=UA, timeout=6)
    return float(r.json()["data"][0]["last"])


def get_candles(symbol=None, interval=None, limit=None):
    symbol = symbol or cfg.SYMBOL
    interval = interval or cfg.TIMEFRAME
    limit = limit or cfg.CANDLES
    r = requests.get(f"{BASE}/v5/market/kline",
                     params={"category": "spot", "symbol": symbol, "interval": _IV.get(interval, "5"), "limit": limit},
                     headers=UA, timeout=8)
    data = r.json()
    if data.get("retCode") != 0:
        raise RuntimeError("Bybit: " + str(data.get("retMsg")))
    return [{"time": int(k[0]), "open": float(k[1]), "high": float(k[2]),
             "low": float(k[3]), "close": float(k[4]), "volume": float(k[5])}
            for k in reversed(data["result"]["list"])]


def get_balance(coin=None):
    coin = coin or cfg.QUOTE_ASSET
    if cfg.PAPER_TRADING:
        return cfg.PAPER_BALANCE
    params = {"accountType": "UNIFIED", "coin": coin}
    r = requests.get(f"{BASE}/v5/account/wallet-balance", params=params,
                     headers=_headers_get(params), timeout=8)
    data = r.json()
    if data.get("retCode") != 0:
        raise RuntimeError("Bybit balance: " + str(data.get("retMsg")))
    lst = data["result"]["list"][0].get("coin", [])
    bal = next((c for c in lst if c["coin"] == coin), None)
    return float(bal["availableToWithdraw"]) if bal and bal.get("availableToWithdraw") else 0.0


def place_order(side, quantity, symbol=None):
    symbol = symbol or cfg.SYMBOL
    if cfg.PAPER_TRADING:
        print(f"[PAPER][BYBIT] {side} {quantity} {symbol}")
        return {"orderId": "PAPER_" + str(int(time.time() * 1000)), "orderStatus": "Filled"}
    body = {"category": "spot", "symbol": symbol, "side": "Buy" if side == "BUY" else "Sell",
            "orderType": "Market", "qty": str(quantity)}
    body_str = json.dumps(body, separators=(",", ":"))
    r = requests.post(f"{BASE}/v5/order/create", data=body_str, headers=_headers_post(body_str), timeout=10)
    data = r.json()
    if data.get("retCode") != 0:
        raise RuntimeError(f"Bybit order failed ({data.get('retCode')}): {data.get('retMsg')}")
    print(f"[LIVE][BYBIT] {side} {quantity} {symbol} -> {data.get('result', {}).get('orderId')}")
    return data["result"]
